asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+    binary_data: Union[BaseGeometry, bytes, str]
+) -> Union[BaseGeometry, None]:
+    """
+    Parse binary geometry data into Shapely geometry object.
+
+    Args:
+        binary_data: Input geometry data in various formats.
+
+    Returns:
+        Shapely geometry object or None if parsing fails.
+    """
+    if isinstance(binary_data, BaseGeometry):
+        return binary_data
+    elif isinstance(binary_data, bytes):
+        return wkb.loads(binary_data)
+    elif isinstance(binary_data, str):
+        return wkb.loads(binary_data, hex=True)
+    else:
+        return None
+
+
+def _parse_capacity(value: Any) -> float:
+    """
+    Parse complex capacity strings and return total capacity as float.
+
+    Args:
+        value: Input capacity value in various formats.
+
+    Returns:
+        Parsed capacity as a float, or NaN if parsing fails.
+    """
+    if pd.isna(value):
+        return np.nan
+
+    if isinstance(value, (int, float)):
+        return float(value)
+
+    value = str(value).strip().lower()
+
+    # Pattern to match "X x Y" format (e.g., "2 x 12.5")
+    pattern = r"(\d+(?:\.\d+)?)\s*x\s*(\d+(?:\.\d+)?)"
+    match = re.match(pattern, value)
+
+    if match:
+        count = float(match.group(1))
+        individual_capacity = float(match.group(2))
+        return count * individual_capacity
+
+    try:
+        return float(value)
+    except ValueError:
+        print(f"Warning: Could not parse capacity value: {value}")
+        return np.nan
+
+
+def process_shapefile(shapefile_directory: str) -> gpd.GeoDataFrame:
+    """
+    Process a shapefile from S3 and return it as a GeoDataFrame.
+
+    Args:
+        shapefile_directory: S3 directory containing the shapefile.
+
+    Returns:
+        GeoDataFrame containing the processed shapefile data.
+    """
+    bucket_name = "asf-heat-pump-suitability"
+    temp_dir = get_shapefile_from_s3(bucket_name, shapefile_directory)
+
+    try:
+        shp_file = next(f for f in os.listdir(temp_dir) if f.endswith(".shp"))
+        shp_path = os.path.join(temp_dir, shp_file)
+        gdf = gpd.read_file(shp_path)
+        gdf = gdf.to_crs(epsg=27700)  # Convert to British National Grid
+        return gdf
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def generate_enw_df() -> gpd.GeoDataFrame:
+    """
+    Generate a GeoDataFrame for Electricity North West (ENW) substations.
+
+    Returns:
+        GeoDataFrame with ENW substation data.
+    """
+    enw_demand = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/dfes-2023-primary-data0.parquet",
+    )
+    enw_substations = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/ndp-pry-bsp-headroom.parquet",
+    )
+    enw_shape = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/ndp-pry-voronoi.parquet",
+    )
+
+    # Filter data for primary substations, current year, and best view scenario
```

Comment:
The substation data for this provider was only available in a forecasting dataset which had different estimated per year on grid enhancements, there were different forecasts, including "Leading the way" (optimistic forecast) and "Falling short" (pessimistic). "Best view" was the middle ground scenario. The forecast was  based on 2023 data so there isn't actually that much variation between the views since it's only a 1 year difference.